docs/proposals/dedicated-nodes.md

```diff
@@ -0,0 +1,210 @@
+NodeConstraint to incoming pods, the value list of the constraint will be very long in clusters
+with lots of dedicated node groups, the admission controller needs to be
+notified whenever a dedicated node group is added or removed or the policy about who
+is allowed to use a dedicated node group changes so that it can label new incoming
+pods correctly *and update pending pods to have the correct NodeConstraint*, etc.
+
+With this in mind, we introduce two new, related concepts to express dedicated node rules:
+*taints* and *tolerations*. In this document we explain how to use them to implement dedicated
+nodes, but they are generic concepts that have applicability beyond dedicated nodes.
+A *taint* is a new type of object that is part of the `NodeSpec`; when present, it prevents pods
+from scheduling on the node unless the pod *tolerates* all of the node's taints. For the purposes
+of this proposal, taints are only associsted with nodes and tolerations are
+only associated with pods, but in theory each could also be associated with other
+types of resources. This document only describes how to use taints and tolerations
+to enforce dedicated nodes; in a followup document we will describe other anticipated uses.
+(Some are alluded to in #17190, the issue on which this proposal is based.)
+
+With taints and tolerations, implementing dedicated nodes is simple: every node that is
+part of a dedicated node group gets a taint of the form `dedicated=<group name>`, and
+every pod that is allowed to use a node in that group gets a toleration of the form
+`dedicated=<group name>`. The vast majority of pods, which are not in any dedicated node
+group, need not be touched, since the taint automatically prevents them from scheduling
+onto the dedicated nodes. This approach easily generalizes to pods that are allowed to
+schedule into multiple dedicated node groups, and nodes that are a member of multiple
+dedicated node groups.
+
+## API
+
+```go
+// The object this Taint is attached to has the effect "effect" on
+// any object that that does not tolerate the Taint.
+type Taint struct {
+	Key string  `json:"key" patchStrategy:"merge" patchMergeKey:"key"`
+	Value string  `json:"value,omitempty"`
+	Effect TaintEffect  `json:"effect"`
+}
+
+// The object this Toleration is attached to tolerates any taint
+// that matches the triple <key,value,effect>.
+type Toleration struct
+	Key string  `json:"key" patchStrategy:"merge" patchMergeKey:"key"`
+	// Wildcard (*) is allowed for value, so that a pod can tolerate all taints
+	// of a particular category. But value is otherwise just a regular string.
+	Value string  `json:"value,omitempty"`
+	Effect TaintEffect  `json:"effect"`
+}
+
+type TaintEffect string
+
+const (
+	// Do not allow new pods to start unless they tolerate the taint,
+	// and allow all already-running pods to continue running.
+	TaintEffectNoStart TaintEffect = "noStart"
```

Comment:
I don't think that the problem you're describing is a big deal. Someone who adds a noStart taint by definition doesn't care about what is running on the machine, they just don't want new stuff to start running there. So I don't think they care what the final state is in the scenario you described.

The main use case for the noStart taint is exactly the one you mentioned -- to "cordon" (#16698) a machine in preparation for draining.